Add global tab width setting & apply use in code editor

Jeehut opened the pull request:
<!--- REQUIRED: Provide a general summary of your changes in the Title above -->

### Description

This adds a new "Global Tab Width" setting to the general settings pane and uses the current value in the code editor.

### Releated Issue

https://github.com/CodeEditApp/CodeEdit/issues/277

### Checklist (for drafts):

<!--- Add things that are not yet implemented above -->
- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] Review requested

### Screenshots (if appropriate):

<img width="712" alt="Default Tab Width Setting" src="https://user-images.githubusercontent.com/6942160/160261948-a5e845bd-cd21-47bb-af18-e50b8e3e3df1.png">

wdg reviewed (approved):
Looks good 😊

RayZhao1998 commented:
It's a little different from what Xcode looks like. It should have a `TextInput` for user to input too.

<img width="942" alt="截屏2022-03-27 下午9 28 55" src="https://user-images.githubusercontent.com/22616933/160283952-5479d745-1207-47f5-9e7e-5bd33e7d5ebe.png">

Jeehut commented:
@RayZhao1998  I had it as a text input field first, but problem with that is that we have to add extra logic for dealing with character entry and invalid numbers like 0 or negatives. I find, given that it's a range of just 6 possible values (2-8), a stepper should be fast enough, even when using VoiceOver. I don't know why Apple is using a text field for that ... I don't think it's the best UX here. But I'm open to other opinions. 🤷‍♂️

austincondiff commented:
What about a select input?

lukepistrol commented:
> What about a select input?

Since it's only a couple of options to choose from I would also suggest using a picker/menu to set this.

Jeehut commented:
As requested, I just changed the stepper UI to a picker UI. I also implemented the related `TabWidth` enum and localized it properly using pluralization (so the options read nicely when scaling to other languages).
<img width="623" alt="Picker" src="https://user-images.githubusercontent.com/6942160/160303155-62ba3904-5f59-41c2-9194-3f4a7ebaaf7e.png">

austincondiff commented:
We probably only need 2 and 4. I don't know if any setup that requires indentation of 3 or 7 spaces and I think 8 is just too much.

austincondiff commented:
I'd like to see an "Indent Using: Tabs | Spaces" select as well (if you consider that in the scope of this PR).

Jeehut commented:
@austincondiff While it's less common, people do actually use other settings than 2 or 4 spaces. For example, for some time I actually opted for 3 spaces myself because I felt it combines the best aspects of 2 spaces (more code in one line) and 4 spaces (better visual distinction), but some people are stubborn about even space numbers so – working in a team – I got used to 2 spaces. I also have met people who are big advocates of really large spacing in their code, and GitHub itself rendered tabs as 8 whitespaces (which caused [this SO question](https://stackoverflow.com/questions/8833953/how-to-change-tab-size-on-github)) for a long time by default.

So I don't think limiting the options to 2 and 4 is a good idea, we should stay a bit more flexible here and should rather have more options than less – it doesn't hurt the UX anyways, so why not stay flexible? We could still remove some specific options like "7" at some point if we see that no one is using it in analytics data or add "10" if people ask for it.

As for the "Indent Using: Tabs | Spaces", for me it's "out of scope" of this PR (my goal was to make the already existing constant number `4` in the codebase configurable).

austincondiff commented:
You have a point. I'm not in the business of enforcing or imposing development practices. So if this is the case, go back to a number input with the stepper controls and cap it at 8. Sorry to have you make so many changes.

Jeehut commented:
@austincondiff Ok sure, I've just reverted my last commit, back to the stepper then. :)

lukepistrol commented on CodeEdit/Settings/GeneralSettingsView.swift:
```diff
@@ -15,6 +15,7 @@ struct GeneralSettingsView: View {
     @AppStorage(ReopenBehavior.storageKey) var reopenBehavior: ReopenBehavior = .default
     @AppStorage(FileIconStyle.storageKey) var fileIconStyle: FileIconStyle = .default
     @AppStorage(CodeFileView.Theme.storageKey) var editorTheme: CodeFileView.Theme = .atelierSavannaAuto
+    @AppStorage("defaultTabWidth") var defaultTabWidth: Int = 4
```
Please don't use String literals as identifiers for `UserDefaults`. In this case just use a case-less `enum` containing the default value and a `storageKey`.

Example: 
```swift
public enum EditorTabWidth {
	static public let `default`: Int = 4 
	static public let storageKey = "defaultTabWidth"
}
```

Jeehut replied:
Good suggestion! Added in cd76e3460356333e35bfb4d48f76025019307198.

lukepistrol commented on CodeEditModules/Modules/CodeFile/src/TextView/CodeEditorTextView.swift:
```diff
@@ -7,8 +7,8 @@
-class CodeEditorTextView: NSTextView {
-    private let tabNumber = 4
+public class CodeEditorTextView: NSTextView {
+    public static var tabWidth = UserDefaults.standard.value(forKey: "defaultTabWidth") as? Int ?? 4
```
Use `@AppStorage` instead

```swift
@AppStorage(EditorTabWidth.storageKey) var tabWidth: Int = EditorTabWidth.default
```

Jeehut replied:
@lukepistrol I'd love to do this, but I can't access `EditorTabWidth` within the `CodeFile` module, the `Settings > Models` folder lies within the app project, I don't think importing the app into a module is a good idea (or even possible). I would need to move the tab width enum to the module then, is that fine with you? Should I even extract the entire "Settings" folder into a module (if possible easily)?

Jeehut replied:
Also, can I please put `@AppStorage(EditorTabWidth.storageKey)` into its own line? The code is so hard to skim through with these `var ...` starting so much to the right, I think all attributes should always be at the start of the line so the code is easier to scan. I don't know why Apple is defaulting to putting it all in one line in their code samples, but I think it's less readable ...

lukepistrol replied:
Moving it to the `CodeFile` module is good for now. @MarcoCarnevali is already working on moving everything to its own module but this takes some time. For now put it into the `CodeFile` module and we will move it to the dedicated `Settings` module once the migration is done!

lukepistrol replied:
> Also, can I please put `@AppStorage(EditorTabWidth.storageKey)` into its own line? The code is so hard to skim through with these `var ...` starting so much to the right, I think all attributes should always be at the start of the line so the code is easier to scan. I don't know why Apple is defaulting to putting it all in one line in their code samples, but I think it's less readable ...

Yeah go ahead but please make sure to change every `@AppStorage` statement so it is the same everywhere

Jeehut replied:
I just added the `EditorTabWidth` to the module instead of the app, this way it worked. See cd76e3460356333e35bfb4d48f76025019307198.

Jeehut commented on CodeEditModules/Modules/CodeFile/src/TextView/CodeEditorTextView.swift:
```diff
@@ -6,9 +6,11 @@
 public class CodeEditorTextView: NSTextView {
-    public static var tabWidth = UserDefaults.standard.value(forKey: "defaultTabWidth") as? Int ?? 4
+    @AppStorage(EditorTabWidth.storageKey)
+    public static var tabWidth: Int = EditorTabWidth.default
```
I'm all in favor of consistency, but I will follow your suggestion from https://github.com/CodeEditApp/CodeEdit/pull/281#discussion_r836482626 instead, I adjusted all property wrappers in the project (adjusting only @AppStorage lead to inconsistencies within types where multiple property wrappers are used).

lukepistrol reviewed (approved):
LGTM
